src/tracking.ts

```diff
@@ -0,0 +1,222 @@
+              })
+                .then(() => {
+                  this.events[value][i].submitted = true;
+                  this.save();
+                })
+                .catch(() => {
+                  console.error(
+                    "Search.io event failed to send.",
+                    this.events[value][i]
+                  );
+                })
+            );
+          }
+        }
+      );
+    });
+
+    return Promise.all(sends).catch(() => {
+      // We already log fails above
+    });
+  }
+
+  async sendEvent(
+    queryId: string,
+    type: string,
+    data: Record<string, string | Metadata>
+  ): Promise<void> {
+    const response = await fetch(
+      `${this.endpoint}/v4/collections/${this.collection}:trackEvent`,
+      {
+        method: "POST",
+        headers: {
+          Accept: "application/json",
+          // XXX: This is to remove the need for the OPTIONS request
+          // https://stackoverflow.com/questions/29954037/why-is-an-options-request-sent-and-can-i-disable-it
+          "Content-Type": "text/plain",
+          "X-Account-Id": this.account,
+        },
+        body: JSON.stringify({
+          query_id: queryId,
+          type,
+          ...data,
+        }),
+      }
+    );
+
+    if (!response.ok) {
+      let message = response.statusText;
+      try {
+        let r = await response.json();
+        if (r.message) {
+          message = r.message;
+        }
+      } catch (_) {}
+
+      if (response.status === 403) {
+        throw new RequestError(
+          response.status,
+          "This domain is not authorized to make this request.",
+          new Error(message)
+        );
+      }
+
+      throw new RequestError(
+        response.status,
+        "Request failed due to a configuration error.",
+        new Error(message)
+      );
+    }
+
+    return response.json();
+  }
+
+  private getExpiry() {
+    return Date.now() - EXPIRY_DAYS * 24 * 60 * 60 * 1000;
+  }
+
+  purge() {
+    const expiry = this.getExpiry();
+    let requiresSave = false;
+
+    Object.keys(this.events).forEach((value) => {
+      const purged = this.events[value].filter(
+        ({ submitted, timestamp }) => !submitted || timestamp > expiry
+      );
+      if (purged.length < this.events[value].length) {
+        this.setEvents(value, purged);
+        requiresSave = true;
+      }
+    });
+    requiresSave && this.save();
+  }
+
+  getEvents(value: Value): EventState[] {
+    return (this.events[value] || []).slice();
+  }
+
+  setEvents(value: Value, events: EventState[]) {
+    if (events.length) {
+      this.events[value] = events;
+    } else {
+      delete this.events[value];
+    }
+  }
+
+  updateQueryID(queryId: string) {
+    this.queryId = queryId;
+  }
+
+  track(type: string, value: Value, metadata?: Metadata) {
+    let queryId = this.queryId;
+    if (!FUNNEL_ENTRY_TYPES.includes(type)) {
+      const lastEvent = this.getEvents(value).reverse().pop();
```

Comment:
I think it'd be clearer to just lookup by index rather than calling two mutation functions.

Reply:
I uncovered a bug in this code due to your comment, the `reverse` mutation was incorrect.